File: python/src/n0235_lowest_common_ancestor_of_a_binary_search_tree_test.py
from n0235_lowest_common_ancestor_of_a_binary_search_tree import lowest_common_ancestor
from utils.tree import Tree


def test_lowest_common_ancestor_example_one() -> None:
    tree = Tree([6, 2, 8, 0, 4, 7, 9, None, None, 3, 5])

    p = next(filter(lambda node: node.value == 2, tree.nodes()))
    q = next(filter(lambda node: node.value == 8, tree.nodes()))
    expected = next(filter(lambda node: node.value == 6, tree.nodes()))

    assert lowest_common_ancestor(tree.root, p, q) == expected


def test_lowest_common_ancestor_example_two() -> None:
    tree = Tree([6, 2, 8, 0, 4, 7, 9, None, None, 3, 5])

    p = next(filter(lambda node: node.value == 2, tree.nodes()))
    q = next(filter(lambda node: node.value == 4, tree.nodes()))
    expected = next(filter(lambda node: node.value == 2, tree.nodes()))

    assert lowest_common_ancestor(tree.root, p, q) == expected


def test_lowest_common_ancestor_example_three() -> None:
    tree = Tree([2, 1])

    p = next(filter(lambda node: node.value == 2, tree.nodes()))
    q = next(filter(lambda node: node.value == 1, tree.nodes()))
    expected = next(filter(lambda node: node.value == 2, tree.nodes()))

    assert lowest_common_ancestor(tree.root, p, q) == expected

File: rust/src/n0235_lowest_common_ancestor_of_a_binary_search_tree.test.rs
use crate::{n0235_lowest_common_ancestor_of_a_binary_search_tree::*, utils::tree::Tree};

#[test]
fn test_lowest_common_ancestor_example_one() {
    let tree = Tree::new([
        Some(6),
        Some(2),
        Some(8),
        Some(0),
        Some(4),
        Some(7),
        Some(9),
        None,
        None,
        Some(3),
        Some(5),
    ]);
    let p = tree.nodes().find(|n| n.borrow().value == 2);
    let q = tree.nodes().find(|n| n.borrow().value == 8);

    let expected = tree.nodes().find(|n| n.borrow().value == 6);
    assert_eq!(lowest_common_ancestor(tree.root, p, q), expected);
}

#[test]
fn test_lowest_common_ancestor_example_two() {
    let tree = Tree::new([
        Some(6),
        Some(2),
        Some(8),
        Some(0),
        Some(4),
        Some(7),
        Some(9),
        None,
        None,
        Some(3),
        Some(5),
    ]);
    let p = tree.nodes().find(|n| n.borrow().value == 2);
    let q = tree.nodes().find(|n| n.borrow().value == 4);

    let expected = tree.nodes().find(|n| n.borrow().value == 2);
    assert_eq!(lowest_common_ancestor(tree.root, p, q), expected);
}

#[test]
fn test_lowest_common_ancestor_example_three() {
    let tree = Tree::new([Some(2), Some(1)]);
    let p = tree.nodes().find(|n| n.borrow().value == 2);
    let q = tree.nodes().find(|n| n.borrow().value == 1);

    let expected = tree.nodes().find(|n| n.borrow().value == 2);
    assert_eq!(lowest_common_ancestor(tree.root, p, q), expected);
}
